### PR TITLE
Feat/tweaks for network update

### PR DESCRIPTION
**Pull request by adekbadek**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

A few tweaks for the upcoming changes to `newspack-network`, and one fix:

1. Extricates a `get_reader_roles` method in Reader Activation
2. Adds metadata to webhook request posts, so they can be found later (and avoid duplication)
3. Fixes an issue where `subscription_id` was set only on renewal orders, not on the initial order of a subscription

### How to test the changes in this Pull Request:

_Follow instructions in https://github.com/Automattic/newspack-network/pull/51_

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: includes/reader-activation/class-reader-activation.php [hunk 1]
```diff
@@ -679,7 +686,7 @@ public static function is_user_reader( $user, $strict = false ) {
 			 *
```
Comment (leogermani) [inline includes/reader-activation/class-reader-activation.php]: Move the filter doc along with the filter

File: includes/data-events/class-utils.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+		];
+	}
+
+	/**
+	 * Get subscription data.
+	 *
+	 * @param WC_Subscription $subscription Subscription.
+	 */
+	public static function get_subscription_data( $subscription ) {
+		$product_id = \Newspack\Donations::get_order_donation_product_id( $subscription->get_id() );
```
Comment (leogermani) [inline includes/data-events/class-utils.php:73]: we are using this for any kind of subscription, so we should not check and return nothing if it's not a donation subscription. I think this check should still be kept in the listerers

Note that we still need to fetch the product ID to be able to get the recurrence

File: includes/data-events/class-utils.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+		$order = \wc_get_order( $order_id );
+		if ( ! $order ) {
+			return;
+		}
+		if ( ! \Newspack\WooCommerce_Connection::should_sync_order( $order ) ) {
+			return;
+		}
+		// Donation orders always have just a single product, but other orders can have more than one.
+		$product_id = \Newspack\Donations::get_order_donation_product_id( $order_id ) ?? array_values( \Newspack\WooCommerce_Connection::get_products_for_order( $order_id ) );
+		if ( ! $product_id ) {
```
Comment (leogermani) [inline includes/data-events/class-utils.php:34]: Make sure it works for any order and not only for donations

File: includes/data-events/listeners.php [hunk 4]
```diff
@@ -134,37 +134,7 @@ function( $provider, $email, $lists_added, $lists_removed, $result ) {
...
-			'referer'         => $order->get_meta( '_newspack_referer' ),
-			'popup_id'        => $order->get_meta( '_newspack_popup_id' ),
-			'is_renewal'      => $is_renewal,
-			'subscription_id' => $subscription_id,
-			'platform_data'   => [
-				'order_id'   => $order_id,
-				'product_id' => $product_id,
-			],
-		];
+		return \Newspack\Data_Events\Utils::get_order_data( $order_id );
```
Comment (leogermani) [inline includes/data-events/listeners.php]: here is where we want to check if this is a donation related order or not

File: includes/data-events/listeners.php [hunk 5]
```diff
@@ -178,35 +148,7 @@ function( $order_id, $product_id ) {
...
-			'popup_id'        => $order->get_meta( '_newspack_popup_id' ),
-			'is_renewal'      => $is_renewal,
-			'subscription_id' => $subscription_id,
-			'platform_data'   => [
-				'order_id'   => $order_id,
-				'product_id' => $product_id,
-				'client_id'  => $order->get_meta( NEWSPACK_CLIENT_ID_COOKIE_NAME ),
-			],
-		];
+		return \Newspack\Data_Events\Utils::get_order_data( $order_id );
```
Comment (leogermani) [inline includes/data-events/listeners.php]: same as above. Here is where we should check if it's a donation.

The event name is misleading btw

File: includes/data-events/listeners.php [hunk 6]
```diff
@@ -246,36 +188,7 @@ function( $subscription, $status_to, $status_from ) {
...
```
Comment (leogermani) [inline includes/data-events/listeners.php]: check donation

File: includes/data-events/listeners.php [hunk 7]
```diff
@@ -325,20 +213,7 @@ function( $subscription, $status_to, $status_from ) {
...
-		return [
-			'user_id'         => $subscription->get_customer_id(),
-			'email'           => $subscription->get_billing_email(),
-			'subscription_id' => $subscription->get_id(),
-			'amount'          => (float) $subscription->get_total(),
-			'currency'        => $subscription->get_currency(),
-			'recurrence'      => empty( $recurrence ) ? 'once' : $recurrence,
-			'platform'        => Donations::get_platform_slug(),
-		];
+		return \Newspack\Data_Events\Utils::get_subscription_data( $subscription );
```
Comment (leogermani) [inline includes/data-events/listeners.php]: again a misleading event name. But the check to see if it's a donation should happen here, not in the Utils method

File: includes/data-events/listeners.php [hunk 8]
```diff
@@ -349,22 +224,17 @@ function( $subscription, $status_to, $status_from ) {
 	'woocommerce_subscription_status_updated',
 	'donation_subscription_changed',
 	function( $subscription, $status_to, $status_from ) {
-		$product_id = Donations::get_order_donation_product_id( $subscription->get_id() );
-		if ( ! $product_id ) {
+		$data = \Newspack\Data_Events\Utils::get_subscription_data( $subscription );
```
Comment (leogermani) [inline includes/data-events/listeners.php]: same

**Review by leogermani — CHANGES_REQUESTED** [review]

File: includes/data-events/class-utils.php (hunk 3 above)
Comment (adekbadek) [reply includes/data-events/class-utils.php:34]: That's what happens here – if the donation product ID is not found, the fallback is `\Newspack\WooCommerce_Connection::get_products_for_order`.

File: includes/data-events/class-utils.php (hunk 2 above)
Comment (adekbadek) [reply includes/data-events/class-utils.php:73]: > we are using this for any kind of subscription

No, this is only for recurring donations (`donation_subscription_changed`, `donation_subscription_cancelled`) listeners. `product_subscription_changed` listener has its own data processing below.

**Comment by adekbadek** [comment]
Addressed the is-donation-order/subscription comments in 85a624b042e0960878e4b683b7d22f3d524e30e2.

File: includes/reader-activation/class-reader-activation.php (hunk 1 above)
Comment (adekbadek) [reply includes/reader-activation/class-reader-activation.php]: Done in 2eb469d

File: includes/data-events/class-utils.php (hunk 2 above)
Comment (leogermani) [reply includes/data-events/class-utils.php:73]: Let's rename these methods then? `...donation_subscriptions...` ? or even `get_recurring_donations_data` ?

Comment (leogermani) [reply includes/data-events/class-utils.php:73]: oops, I commented before looking at the changes :)

**Review by leogermani — APPROVED** [review]

**Pull request merged**